File: dos/common/protocol.c
#include "protocol.h"

unsigned short ime_crc16(const unsigned char *data, unsigned short length)
{
    unsigned short crc = 0xFFFF;
    unsigned char bit;

    while (length--) {
        crc ^= *data++;
        for (bit = 0; bit < 8; ++bit)
            crc = (crc & 1) ? (crc >> 1) ^ 0xA001 : crc >> 1;
    }
    return crc;
}

unsigned short ime_packet_build(unsigned char *out, unsigned char type,
                                unsigned short sequence,
                                const unsigned char *payload,
                                unsigned short payload_length)
{
    unsigned short i;
    unsigned short crc;

    if (payload_length > IME_MAX_PAYLOAD)
        return 0;
    out[0] = IME_MAGIC0;
    out[1] = IME_MAGIC1;
    out[2] = IME_VERSION;
    out[3] = type;
    out[4] = (unsigned char)sequence;
    out[5] = (unsigned char)(sequence >> 8);
    out[6] = (unsigned char)payload_length;
    out[7] = (unsigned char)(payload_length >> 8);
    for (i = 0; i < payload_length; ++i)
        out[8 + i] = payload[i];
    crc = ime_crc16(out, 8 + payload_length);
    out[8 + payload_length] = (unsigned char)crc;
    out[9 + payload_length] = (unsigned char)(crc >> 8);
    return 10 + payload_length;
}

int ime_packet_parse(const unsigned char *packet, unsigned short packet_length,
                     unsigned char *type, unsigned short *sequence,
                     const unsigned char **payload,
                     unsigned short *payload_length)
{
    unsigned short length;
    unsigned short expected;
    unsigned short actual;

    if (packet_length < 10 || packet[0] != IME_MAGIC0 || packet[1] != IME_MAGIC1 ||
        packet[2] != IME_VERSION)
        return 0;
    length = (unsigned short)packet[6] | ((unsigned short)packet[7] << 8);
    expected = 10 + length;
    if (length > IME_MAX_PAYLOAD || packet_length != expected)
        return 0;
    actual = (unsigned short)packet[8 + length] |
             ((unsigned short)packet[9 + length] << 8);
    if (actual != ime_crc16(packet, 8 + length))
        return 0;
    *type = packet[3];
    *sequence = (unsigned short)packet[4] | ((unsigned short)packet[5] << 8);
    *payload = packet + 8;
    *payload_length = length;
    return 1;
}

File: tools/protocol.py
"""Wire codec for the DOS IME Bridge protocol."""

from __future__ import annotations

import struct
from dataclasses import dataclass

MAGIC = b"ID"
VERSION = 1
MAX_PAYLOAD = 512
HEADER = struct.Struct("<2sBBHH")
CRC = struct.Struct("<H")

HELLO = 1
PING = 2
PONG = 3
OPEN_IME = 4
TEXT = 5
TEXT_ACK = 6


def crc16(data: bytes) -> int:
    value = 0xFFFF
    for byte in data:
        value ^= byte
        for _ in range(8):
            value = (value >> 1) ^ 0xA001 if value & 1 else value >> 1
    return value


@dataclass(frozen=True)
class Packet:
    kind: int
    sequence: int
    payload: bytes = b""

    def encode(self) -> bytes:
        if not 0 <= self.kind <= 255:
            raise ValueError("packet type out of range")
        if not 0 <= self.sequence <= 0xFFFF:
            raise ValueError("sequence out of range")
        if len(self.payload) > MAX_PAYLOAD:
            raise ValueError("payload exceeds negotiated maximum")
        header = HEADER.pack(MAGIC, VERSION, self.kind, self.sequence, len(self.payload))
        return header + self.payload + CRC.pack(crc16(header + self.payload))


def decode(data: bytes) -> Packet:
    if len(data) < HEADER.size + CRC.size:
        raise ValueError("short packet")
    magic, version, kind, sequence, length = HEADER.unpack_from(data)
    if magic != MAGIC or version != VERSION:
        raise ValueError("invalid packet header")
    if length > MAX_PAYLOAD or len(data) != HEADER.size + length + CRC.size:
        raise ValueError("invalid packet length")
    payload = data[HEADER.size : HEADER.size + length]
    actual = CRC.unpack_from(data, HEADER.size + length)[0]
    if actual != crc16(data[: HEADER.size + length]):
        raise ValueError("CRC mismatch")
    return Packet(kind, sequence, payload)


class StreamDecoder:
    def __init__(self) -> None:
        self.buffer = bytearray()

    def feed(self, data: bytes) -> list[Packet]:
        self.buffer.extend(data)
        packets: list[Packet] = []
        while len(self.buffer) >= HEADER.size + CRC.size:
            if self.buffer[:2] != MAGIC:
                del self.buffer[0]
                continue
            length = struct.unpack_from("<H", self.buffer, 6)[0]
            total = HEADER.size + length + CRC.size
            if length > MAX_PAYLOAD:
                del self.buffer[0]
                continue
            if len(self.buffer) < total:
                break
            candidate = bytes(self.buffer[:total])
            try:
                packets.append(decode(candidate))
            except ValueError:
                del self.buffer[0]
                continue
            del self.buffer[:total]
        return packets
